src/vivarium/framework/lifecycle.py

```diff
@@ -193,80 +351,233 @@ def _wrapped(*args, **kwargs):
+        state_name
+            The name of the state to register the handlers for.
+        handlers
+            A list of functions that will execute during the state.
+
+        """
         s = self.lifecycle.get_state(state_name)
         s.add_handlers(handlers)
 
-    def add_constraint(self, constrained_function, allow_during=None, restrict_during=None):
+    def add_constraint(self, method: MethodType,
+                       allow_during: List[str] = (),
+                       restrict_during: List[str] = ()):
+        """Constrains a function to be executable only during certain states.
+
+        Parameters
+        ----------
+        method
+            The method to add constraints to.
+        allow_during
+            An optional list of life cycle states in which the provided
+            method is allowed to be called.
+        restrict_during
+            An optional list of life cycle states in which the provided
+            method is restricted from being called.
+
+        Raises
+        ------
+        ValueError
+            If neither ``allow_during`` nor ``restrict_during`` are provided,
+            or if both are provided.
+        LifeCycleError
+            If states provided as arguments are not in the life cycle.
+
+        """
         if allow_during and restrict_during or not (allow_during or restrict_during):
             raise ValueError('Must provide exactly one of "allow_during" or "restrict_during".')
+        unknown_states = set(allow_during).union(restrict_during).difference(self.lifecycle._state_names)
+        if unknown_states:
+            raise LifeCycleError(f'Attempting to constrain {method} with '
+                                 f'states not in the life cycle: {list(unknown_states)}.')
+        if id(method) in self._constraints:
+            raise ConstraintError(f'Method {method} has already been constrained.')
+
         if restrict_during:
             allow_during = [s for s in self.lifecycle._state_names if s not in restrict_during]
-        unknown_states = set(allow_during).difference(self.lifecycle._state_names)
-        if unknown_states:
-            raise LifeCycleError(f'Attempting to constrain {constrained_function} to '
-                                 f'unknown states {list(unknown_states)}.')
-        self._make_constraint(constrained_function, allow_during)
 
-    def __repr__(self):
+        constrained_method = self._make_constraint(method, allow_during)
+        self._constraints.update({id(method), id(constrained_method)})
+
+    def __repr__(self) -> str:
         return f'LifeCycleManager(state={self.current_state})'
 
-    def __str__(self):
+    def __str__(self) -> str:
         return str(self.lifecycle)
 
 
 class LifeCycleInterface:
+    """Interface to the life cycle management system.
+
+    The life cycle management system allows components to constrain
+    methods so that they're only available during certain simulation
+    life cycle states.
```

Comment:
maybe also a line about ordering here?

Reply:
Ordering contracts are handled privately in the simulation context.  If a user ever sees one, we're in trouble.